x-pack/plugins/ml/server/lib/capabilities/check_capabilities.ts

```diff
@@ -61,11 +62,16 @@ function disableAdminPrivileges(capabilities: MlCapabilities) {
 export type HasMlCapabilities = (capabilities: MlCapabilitiesKey[]) => Promise<void>;
 
 export function hasMlCapabilitiesProvider(resolveMlCapabilities: ResolveMlCapabilities) {
+  const resolveMlCapabilitiesMemo = memoize(
+    async (request: KibanaRequest) => await resolveMlCapabilities(request),
```

Comment:
is this not a memory leak?

Reply:
Ok thanks, that sounds good. What about creating a function and wrapping it with `_.once`? Maybe a little cleaner than storing a promise in a variable?

Reply:
Btw, can you use `router.registerRequestContext` to provide a shared function that returns capabilities?

Reply:
I can't find any reference to `registerRequestContext` in kibana? Can you share an example?

Reply:
Ah sorry, it's `core.http.registerRouteHandlerContext`.

Reply:
e.g.: https://github.com/elastic/kibana/blob/main/x-pack/plugins/licensing/server/plugin.ts#L129-L132

Reply:
ah ok, it looks like this is just for customising the context for the router, rather than for functions we'd share out where we need the request object passed in.
But to be honest, our shared functions and their providers were written so long ago that I imagine there are better ways to do this now. I'd love to get rid of the providers which require the request object passed in.

Reply:
> ah ok, it looks like this is just for customising the context for the router, rather than for functions we'd share out where we need the request object passed in.

Can you clarify what you mean here? Specifically "just for customising the context for the router".

Reply:
Unless I'm misunderstanding what you were suggesting, context created by the `registerRouteHandlerContext` is available inside the route handlers.
However the functions we're sharing with other plugins via our "provider" functions are not inside route handlers.